san/impl/derive_type.cpp

```diff
@@ -0,0 +1,256 @@
+#include <iostream>
+
+#include "san.hpp"
+namespace san {
+
+using namespace ast;
+
+void DeriveType::visit(Program* node) {
+
+    // check variables with both type and initial value
+    for (auto var : node->variables) {
+        var->accept(*this);
+    }
+    // to check array length type
+    for (auto type : node->types) {
+        type->type->accept(*this);
+    }
+    // routine has an expression
+    for (auto routine : node->routines) {
+        routine->accept(*this);
+    }
+}
+
+void DeriveType::visit(RoutineDecl* node) {
+    m_inRoutineParams = true;
+    for (auto parameter : node->parameters) {
+        parameter->accept(*this);
+    }
+    m_inRoutineParams = false;
+    node->body->accept(*this);
+}
+
+void DeriveType::visit(Type*) {}
+
+void DeriveType::visit(AliasedType* node) { node->actualType->accept(*this); }
+
+void DeriveType::visit(PrimitiveType*) {}
+
+void DeriveType::visit(IntegerType*) {}
+
+void DeriveType::visit(RealType*) {}
+
+void DeriveType::visit(BooleanType*) {}
+
+void DeriveType::visit(ArrayType* node) {
+
+    if (node->length) {
+        node->length->accept(*this);
+        // any primitive type can be convered to the int
+        if (!typeIsBase(node->length->type)) {
+            error(node->begin, "invalid array length type, should be integer");
+        }
+    } else if (!m_inRoutineParams) {
+        error(node->begin, "length of array should be always defined");
+    } // length can be absent in case of routine parameter
+    node->elementType->accept(*this);
+    if (m_searchArray) {
+        m_arrayInnerType = node->elementType;
+    }
+}
+
+void DeriveType::visit(RecordType* node) {
+    for (auto field : node->fields) {
+        field->accept(*this);
+        if (m_searchRecord && field->name == m_recordField) {
+            m_recordInnerType = field->type;
+        }
+    }
+}
+
+void DeriveType::visit(VariableDecl* node) {
+
+    // check type
+    if (node->type != nullptr) {
+        node->type->accept(*this);
+    }
+    // check/evaluate expression value
+    if (node->initialValue != nullptr) {
+        node->initialValue->accept(*this);
+    }
+    // assign initial value
+    if (node->type == nullptr) {
+        node->type = node->initialValue->type;
+    }
+}
+
+void DeriveType::visit(TypeDecl* node) { node->type->accept(*this); }
+
+void DeriveType::visit(Body* node) {
+
+    // to check array length type
+    for (auto type : node->types) {
+        type->type->accept(*this);
+    }
+    // check variables with both type and initial value
+    for (auto var : node->variables) {
+        var->accept(*this);
+    }
+    // routine has an expression
+    for (auto statement : node->statements) {
+        statement->accept(*this);
+    }
+}
+
+void DeriveType::visit(Statement*) {}
+
+void DeriveType::visit(ReturnStatement*) {}
+
+void DeriveType::visit(Assignment* node) { node->rhs->accept(*this); }
```

Comment:
The LHS may also need to be checked since it not always a simple identifier that gets linked by the `IdentifierResolver`. We need to link the `a.x` expression with the variable declaration inside the `a` record